toxcore/group.c

```diff
@@ -187,27 +240,51 @@ static int get_group_num(const Group_Chats *g_c, const uint8_t *identifier)
  * return peer number if peer is in chat.
  * return -1 if peer is not in chat.
  *
- * TODO(irungentoo): make this more efficient.
  */
-static int get_peer_index(Group_c *g, uint16_t peer_number)
+static aint get_peer_index(const Group_c *g, uint16_t peer_gid)
 {
-    uint32_t i;
+    aint i;
 
     for (i = 0; i < g->numpeers; ++i) {
-        if (g->group[i].peer_number == peer_number) {
+        if (g->peers[i].gid == (int)peer_gid) {
             return i;
         }
     }
 
     return -1;
 }
 
+static uint16_t find_new_peer_gid(const Group_c *g)
+{
+    uint16_t peer_number;
+    randombytes((uint8_t *)&peer_number, 2);
+
+    size_t tries = 0;
+
+    while (get_peer_index(g, peer_number) != -1) {
+        randombytes((uint8_t *)&peer_number, 2);
+        ++tries;
+
+        if (tries > 32) {
+            /* just find first unused peer number */
+            peer_number = 0;
```

Comment:
Do you understand, what this code do? Do you understand, how gid's are computed?
Ok, I'll try to explain. gid (or peer_num in old code) is just unique 16 bit identifier of peer in current groupchat. For compatibility reason, algorithm of gid generation wasn't changed. And it was just random value for every just joined peer. It was normal for old groupchats, because there is low probability of collision. But this groupchats allow to join one cloud of peers to another cloud of peers, so probability of gid collision is high. There some techniques to reslove collision.
So, what about 32 tries and linear search of free gid number. It is just improvement of old algorithm. Think, what happens when number of peers is about to maximum (64k)? Right, 32 tries to find free gid value can be not enough. And linear search is last chance remedy to find it. Now, do you understand your suggestion go search from current random numbers is wrong?